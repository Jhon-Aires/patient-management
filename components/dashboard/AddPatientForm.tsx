'use client';

import { zodResolver } from '@hookform/resolvers/zod';
import { SubmitHandler, useForm } from 'react-hook-form';
import { z } from 'zod';

import { Button } from '@/components/ui/button';
import { DialogFooter } from '@/components/ui/dialog';
import {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormLabel,
  FormMessage
} from '@/components/ui/form';
import { formFields, initialValues, schema } from '@/lib/forms/addPatientForm';
import { Patient } from '@/types/Patients';

export default function AddPatientForm({
  onSave,
  onCancel
}: {
  onSave: (patient: Patient) => void;
  onCancel: () => void;
}) {
  const form = useForm<z.infer<typeof schema>>({
    resolver: zodResolver(schema),
    defaultValues: { ...initialValues },
    reValidateMode: 'onSubmit'
  });

  const onSubmit: SubmitHandler<typeof initialValues> = ({
    avatar,
    description,
    name
  }) => {
    onSave({
      avatar,
      description,
      name,
      id: 1,
      createdAt: new Date().toISOString(),
      website: ''
    });
  };

  return (
    <Form {...form}>
      <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-4">
        {formFields.map(
          ({
            controller: Controller,
            controllerAttrs,
            label,
            name,
            description
          }) => {
            console.log('controllerAttrs', controllerAttrs);

            return (
              <FormField
                key={name}
                control={form.control}
                name={name as keyof typeof initialValues}
                render={({ field }) => {
                  console.log('controllerAttrs', controllerAttrs);

                  return (
                    <FormItem>
                      {label && <FormLabel>{label}</FormLabel>}
                      <FormControl>
                        <Controller
                          {...field}
                          {...controllerAttrs}
                          className={controllerAttrs?.className}
                        />
                      </FormControl>
                      {description && (
                        <FormDescription>{description}</FormDescription>
                      )}
                      <FormMessage />
                    </FormItem>
                  );
                }}
              />
            );
          }
        )}
        <DialogFooter>
          <Button type="button" variant="outline" onClick={onCancel}>
            Cancel
          </Button>
          <Button type="submit">Add</Button>
        </DialogFooter>
      </form>
    </Form>
  );
}
