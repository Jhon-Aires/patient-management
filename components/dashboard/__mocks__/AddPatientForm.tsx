export default function AddPatientForm() {
  return <div>AddPatientForm</div>;
}
