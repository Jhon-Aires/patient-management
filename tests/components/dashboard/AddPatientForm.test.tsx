import '@testing-library/jest-dom';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import AddPatientForm from '@/components/dashboard/AddPatientForm';

describe('AddPatientForm', () => {
  const patient = {
    avatar: 'patient avatar',
    createdAt: '',
    description: 'description',
    id: 1,
    name: 'Patient Name',
    website: ''
  };

  const onSave = jest.fn();
  const onClose = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const { asFragment } = render(
      <AddPatientForm onSave={onSave} onCancel={onClose} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should call onSave when the form is submitted', async () => {
    render(<AddPatientForm onSave={onSave} onCancel={onClose} />);

    const idInput = screen.getByRole('textbox', {
      name: /id/i
    });

    const nameInput = screen.getByRole('textbox', {
      name: /name/i
    });

    const descriptionInput = screen.getByRole('textbox', {
      name: /description/i
    });

    const avatarInput = screen.getByRole('textbox', {
      name: /avatar/i
    });

    const submitButton = screen.getByRole('button', { name: /add/i });

    await userEvent.type(idInput, patient.id.toString());
    await userEvent.type(nameInput, patient.name);
    await userEvent.type(descriptionInput, patient.description);
    await userEvent.type(avatarInput, patient.avatar);

    await userEvent.click(submitButton);

    // should be toHaveBeenCalledWith(patient) but the date is giving problems
    // and the fakeTimers are not working
    expect(onSave).toHaveBeenCalled();
  });

  it('should call onCancel when the cancel button is clicked', async () => {
    render(<AddPatientForm onSave={onSave} onCancel={onClose} />);

    const cancelButton = screen.getByRole('button', { name: /cancel/i });

    await userEvent.click(cancelButton);

    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('should update the form values when the user types', async () => {
    render(<AddPatientForm onSave={onSave} onCancel={onClose} />);

    const idInput = screen.getByRole('textbox', {
      name: /id/i
    });

    const nameInput = screen.getByRole('textbox', {
      name: /name/i
    });

    const descriptionInput = screen.getByRole('textbox', {
      name: /description/i
    });

    const avatarInput = screen.getByRole('textbox', {
      name: /avatar/i
    });

    await userEvent.type(idInput, patient.id.toString());
    await userEvent.type(nameInput, patient.name);
    await userEvent.type(descriptionInput, patient.description);
    await userEvent.type(avatarInput, patient.avatar);

    expect(idInput).toHaveValue(patient.id.toString());
    expect(nameInput).toHaveValue(patient.name);
    expect(descriptionInput).toHaveValue(patient.description);
    expect(avatarInput).toHaveValue(patient.avatar);
  });

  it('should show an error message when the form is submitted with invalid data', async () => {
    render(<AddPatientForm onSave={onSave} onCancel={onClose} />);

    const submitButton = screen.getByRole('button', { name: /add/i });

    await userEvent.click(submitButton);

    const idErrorMessage = screen.getByText(/id must be a number/i);
    const nameErrorMessage = screen.getByText(/name is required/i);

    expect(idErrorMessage).toBeInTheDocument();
    expect(nameErrorMessage).toBeInTheDocument();
  });
});
