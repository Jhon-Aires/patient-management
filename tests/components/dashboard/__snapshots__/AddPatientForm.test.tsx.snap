// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddPatientForm should render correctly 1`] = `
<DocumentFragment>
  <form
    class="space-y-4"
  >
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r0:-form-item"
      >
        ID
      </label>
      <input
        aria-describedby=":r0:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r0:-form-item"
        name="id"
        placeholder="12345678"
      />
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r1:-form-item"
      >
        Name
      </label>
      <input
        aria-describedby=":r1:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r1:-form-item"
        name="name"
        placeholder="Name"
        value=""
      />
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r2:-form-item"
      >
        Description
      </label>
      <textarea
        aria-describedby=":r2:-form-item-description"
        aria-invalid="false"
        class="flex min-h-[80px] w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 !max-h-80 !min-h-40"
        id=":r2:-form-item"
        name="description"
        placeholder="Description"
      />
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r3:-form-item"
      >
        Avatar
      </label>
      <input
        aria-describedby=":r3:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r3:-form-item"
        name="avatar"
        value=""
      />
    </div>
    <div
      class="flex flex-col-reverse sm:flex-row sm:justify-end sm:space-x-2"
    >
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
        type="button"
      >
        Cancel
      </button>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
        type="submit"
      >
        Add
      </button>
    </div>
  </form>
</DocumentFragment>
`;
